### DISCUSSION TITLE
A third potential inheritance mapping strategy... is it valid?

### DISCUSSION THREAD

**Discussion by gravitiq-cm** [doctrine/doctrine2#11025] [General]
We have cases where we want to use inheritance when defining entities purely because there are common fields or functions for several entities. However, we do not want to use STI (where everything goes into one table) or CTI (where there are entries in multiple tables for a single entity). We want each child class to have its own table, which represents all the fields of that entity, no matter where they came from in the inheritance chain.

To me, this seems to give the benefits of inheritance and code reuse without the downsides of STI/CTI. Is there some reason why this is not considered a valid inheritance mapping strategy and is not supported in Doctrine?

**Comment by greg0ire** [comment]
Isn't that allowed with mapped superclasses?

**Reply by gravitiq-cm** [reply]
My understanding of the release notes for 3.0 is that it will not be possible any more:

> **BC BREAK: Undeclared entity inheritance now throws a MappingException**
> As soon as an entity class inherits from another entity class, inheritance has to
> be declared by adding the appropriate configuration for the root entity.
> 

I read this as you would need to choose STI or CTI and cannot just use "plain inheritance" from 3.0 onwards. Have I misunderstood the changes?

**Reply by greg0ire** [reply]
I think you did. Otherwise `MappedSuperclass` would not longer exist, and it very much does: https://github.com/doctrine/orm/blob/3.0.x/lib/Doctrine/ORM/Mapping/MappedSuperclass.php

**Reply by greg0ire** [reply]
Plus you can see that the deprecation is only emitted when not in the case of a mapped superclass: https://github.com/doctrine/orm/pull/10431/files

**Reply by gravitiq-cm** [reply]
OK, then maybe the docs need some tweaking? It reads here:

> [Entity Inheritance](https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/inheritance-mapping.html#entity-inheritance)
> As soon as one entity class inherits from another entity class, either directly, with a mapped superclass or other unmapped (also called transient) classes in between, these entities form an inheritance hierarchy. The topmost entity class in this hierarchy is called the root entity, and the hierarchy includes all entities that are descendants of this root entity.
> 
> On the root entity class, #[InheritanceType], #[DiscriminatorColumn] and #[DiscriminatorMap] must be specified.
> #[InheritanceType] specifies one of the two available inheritance mapping strategies that are explained in the following sections.
> 

To me that reads like you have to use one of the InheritanceTypes mentioned: either CTI or STI. The third way (neither of these, just using the parent class in an inheritance chain and having a single table for each child entity) is not documented and not mentioned as an InheritanceType.

**Reply by greg0ire** [reply]
I agree that it's confusing… [mapped superclasses](https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/inheritance-mapping.html#mapped-superclasses) are documented just above.

I think everything inside and below the "entity inheritance" paragraph is about inheritance (direct or indirect) between 2 classes tagged with `#[Entity]`. It's not the case when using mapped superclasses, as you can see… there is PHP inheritance, but no Entity inheritance.

If you want to send a PR that changes https://github.com/doctrine/orm/blob/2.16.x/docs/en/reference/inheritance-mapping.rst, I think the first thing to do would be to change the headings so that STI and CTI are under Entity Inheritance.

Another thing you could contribute would be expanding the very first paragraph so that it mentions the distinction between the usage mapped superclasses, that involves PHP inheritance and nothing more, and entity inheritance.

**Reply by gravitiq-cm** [reply]
Wait... are you saying that the Entity Inheritance part does NOT apply to Mapped Superclasses?

I've been using Doctrine for a very long time (>10 years?) and always thought the Entity Inheritance part was linked to Mapped Superclasses as well (it was just more detail for the same concept).

The first sentence in the Entity Inheritance section even calls this out by mentioning Mapped Superclasses!

To paraphrase the first two paragraphs, it reads to me as "when entities inherit from other classes (inc Mapped Superclass) they form an inheritance hierarchy. The top entity is the root entity. On the root entity the InheritanceType (and other things) must be specified. [ie: you should declare InheritanceType for Mapped Superclasses and, further: if you're currently not declaring these things you're hacking around the edges of what's "correct" and may have unforeseen problems]"

**Reply by greg0ire** [reply]
I think your paraphrasing is plain wrong. Instead of paraphrasing anything, let me quote the docs with a little extra emphasis

> As soon as one entity class inherits from another _entity_ class, either directly, with a mapped superclass

So the top entity has to be... well... an entity. And a mapped superclass is not an entity AFAIK.

**Reply by gravitiq-cm** [reply]
One little word makes so much difference 😅

Thanks for clarifying! (and pointing out where I went wrong)

Would be great to find a way to clarify those docs, since there may be others misreading them the way I did. It _does_ say on page that a Mapped Superclass is not an entity, but it wouldn't hurt to be clearer in the Entity Inheritance section. That sentence does continue to say "As soon as one entity class inherits from another entity class, either directly, **with a mapped superclass**..." so it's easy to miss that it only applies when there's a concrete Entity in the chain (as opposed to just extending a Mapped Superclass).

**Reply by greg0ire** [reply]
I agree… would you mind sending a pull request?

**Comment by derrabus** [comment]
> we want to use inheritance when defining entities purely because there are common fields or functions for several entities.

Traits. ✌️

**Reply by gravitiq-cm** [reply]
https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/limitations-and-known-issues.html#using-traits-in-entity-classes

**Reply by gravitiq-cm** [reply]
Basically says "don't use traits to do this" [the scenario I described above]

**Reply by derrabus** [reply]
We should probably revisit that part of the docs. They should solve your problem and if they don't work in your scenario, we should make them work.

**Comment by beberlei** [comment]
The concept exists https://martinfowler.com/eaaCatalog/concreteTableInheritance.html

Implementation is also in Hibernate.

Its complicated, because it requires unions for querying the base tables and id generation is "tricky", because it sort of requires UUIDs or sequences.

**Reply by gravitiq-cm** [reply]
I'm not sure why it's complicated... there's a single table for each entity, so it manages its own IDs and sequences (as if it's not in an inheritance chain at all). In other words, the inheritance only exists in PHP-space, but not in the underlying database.

**Reply by beberlei** [reply]
If each table uses the same id for an entity, say ID 1 and you query for the base class, then you get 3 entities with the same ID from the same inheritance hierachy, leading to all sorts of "identity crisis" issues, notably in the code of Doctrine identity map for example.

Granted, you can find solutions for these problems, its just nobody bothered to implement this yet.
